src/simulator.py

```diff
@@ -433,13 +435,76 @@ def read_profile(ref_g, ref_t, number, model_prefix, per, mode, strandness, exp,
             kde_aligned_2d = joblib.load(model_prefix + "_aligned_region_2d.pkl")
 
 
-def collapse_homo(seq, k):
-    read = re.sub("A" * k + "+", "A" * (k - 1), seq)
-    read = re.sub("C" * k + "+", "C" * (k - 1), read)
-    read = re.sub("T" * k + "+", "T" * (k - 1), read)
-    read = re.sub("G" * k + "+", "G" * (k - 1), read)
+def mutate_homo(seq, k, basecaller):
+    hp_arr = []  # [[base, start, end], ...]
+    hp_length_hist = {}  # {length: {A/T: count, C/G: count} ...}
+    hp_samples = {}  # {length: {A/T: [sample], C/G: [sample]} ...}
 
-    return read
+    # Finding homopolymers in sequence
+    pattern = "A{" + re.escape(str(k)) + ",}|C{" + re.escape(str(k)) + ",}|G{" + re.escape(
+        str(k)) + ",}|T{" + re.escape(str(k)) + ",}"
+
+    for match in re.finditer(pattern, seq):
+        hp_start = match.start()
+        hp_end = match.end()
+        length = hp_end - hp_start
+        base = match.group()[0]
+        hp_arr.append([base, hp_start, hp_end])
+        if length not in hp_length_hist.keys():
+            hp_length_hist[length] = {"A/T": 0, "C/G": 0}
+
+        if base == "A" or base == "T":
+            key = "A/T"
+        else:
+            key = "C/G"
+
+        hp_length_hist[length][key] += 1
+
+    # Obtaining samples from normal distributions
+    for length in hp_length_hist.keys():
+        hp_samples[length] = {}
+        at_mu, at_sigma, cg_mu, cg_sigma = nd.get_nd_par(length, basecaller)
+
+        if hp_length_hist[length]["A/T"] > 0:
+            hp_samples[length]["A/T"] = np.random.normal(at_mu, at_sigma, hp_length_hist[length]["A/T"])
+
+        if hp_length_hist[length]["C/G"] > 0:
+            hp_samples[length]["C/G"] = np.random.normal(cg_mu, cg_sigma, hp_length_hist[length]["C/G"])
+
+    # Mutating homopolymers in given sequence
+    last_pos = 0
+    mutated_seq = ""
+    total_hp_size_change = 0
+    for hp_info in hp_arr:
+        base = hp_info[0]
+        ref_hp_start = hp_info[1]
+        ref_hp_end = hp_info[2]
+
+        if base == "A" or base == "T":
+            size = round(hp_samples[ref_hp_end - ref_hp_start]["A/T"][-1])
+            hp_samples[ref_hp_end - ref_hp_start]["A/T"] = hp_samples[ref_hp_end - ref_hp_start]["A/T"][:-1]
+        else:
+            size = round(hp_samples[ref_hp_end - ref_hp_start]["C/G"][-1])
+            hp_samples[ref_hp_end - ref_hp_start]["C/G"] = hp_samples[ref_hp_end - ref_hp_start]["C/G"][:-1]
+
+        mutated_hp = base * int(size)
+        mutated_hp_with_mis = ""
+        for old_base in mutated_hp:
+            p = random.random()
+            if 0 < p <= 0.02215:
```

Comment:
@theottlo is this number basecaller dependent? If so, can we also put it into nd file?

Reply:
Sure I can put it there! That value (the mismatch rate in homopolymers) differs slightly between basecallers--0.02204, 0.02166, and 0.02215.

Reply:
It seems that I can't approve my own PR... so once you finish making the two changes, plz squash and merge, and go ahead with the benchmarking for the ms. Then I'll work on the release